### PR TITLE
Implement Delegated Authentication feature

### PR DESCRIPTION
**Pull request by tkuntubayev**
## Description
Implemented Delegated Authentication feature based on Adyen Authentication SDK. The solution provided as opt-in which means that the merchant will need to add Adyen Authentication SDK in their project in addition to Drop-In/Component SDK, otherwise DA output will not be generated. By default if DA cannot be performed at any step - the flow will fallback to 3DS2. Internal documentation with all integration steps is available [here](https://hub.is.adyen.com/our-solution/enhancements/3d-secure-20/product-specifics/2-delegated-authentication/android-documentation/).

## Checklist
- [x] Changes are tested manually

Two related focals:
1. FOC-67855
2. FOC-67865

### CODE REVIEW

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/ui/DARegistrationView.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    private val binding: ViewDaRegistrationBinding =
+        ViewDaRegistrationBinding.inflate(LayoutInflater.from(context), this)
+
+    private lateinit var localizedContext: Context
+
+    init {
+        orientation = VERTICAL
+    }
+
+    override fun initView(delegate: ComponentDelegate, coroutineScope: CoroutineScope, localizedContext: Context) {
```
Comment (OscarSpruit) [inline 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/ui/DARegistrationView.kt:46]: Can we also split this in more methods?

File: 3ds2/src/main/res/layout/view_da_authentication.xml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    android:layout_height="wrap_content"
+    android:orientation="vertical"
+    tools:parentTag="android.widget.LinearLayout">
+
+    <ImageView
+        android:id="@+id/ivLogo"
+        android:layout_width="wrap_content"
+        android:layout_height="40dp"
+        android:layout_gravity="center"
+        android:layout_marginTop="44dp"
```
Comment (OscarSpruit) [inline 3ds2/src/main/res/layout/view_da_authentication.xml:15]: All styling should be extracted from the views into styles, so merchants can customise it. Check other views for examples

File: 3ds2/src/main/res/layout/view_da_authentication.xml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+<merge xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:orientation="vertical"
+    tools:parentTag="android.widget.LinearLayout">
+
+    <ImageView
+        android:id="@+id/ivLogo"
```
Comment (OscarSpruit) [inline 3ds2/src/main/res/layout/view_da_authentication.xml:11]: Can you use this convention `imageView_logo` for consistency?

File: 3ds2/src/main/res/layout/view_da_authentication.xml [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:layout_marginHorizontal="24dp"
+        android:layout_marginTop="8dp"
+        android:gravity="center"
+        android:textSize="14sp"
+        tools:text="@string/checkout_3ds2_da_authentication_description" />
+
+    <androidx.constraintlayout.widget.ConstraintLayout
```
Comment (OscarSpruit) [inline 3ds2/src/main/res/layout/view_da_authentication.xml:42]: Why is this wrapper needed? We're doing the same in `QrCodeView`, so it should be working here as well

File: 3ds2/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="checkout_3ds2_da_registration_title">Safe and swift checkout!</string>
```
Comment (OscarSpruit) [inline 3ds2/src/main/res/values/strings.xml:3]: Do we have translations for these?

File: 3ds2/src/main/res/values/styles.xml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  ~ Copyright (c) 2023 Adyen N.V.
+  ~
+  ~ This file is open source and available under the MIT license. See the LICENSE file for more info.
+  ~
+  ~ Created by temirlan on 13/1/2023.
+  -->
+
+<resources>
+
+    <style name="DAMaterialAlertDialogTheme">
```
Comment (OscarSpruit) [inline 3ds2/src/main/res/values/styles.xml:11]: If we can use `AlertDialog.Builder` instead, this shouldn't be needed anymore

File: 3ds2/src/test/java/com/adyen/checkout/adyen3ds2/DefaultAdyen3DS2DelegateTest.kt [hunk 7]
```diff
@@ -97,8 +99,13 @@ internal class DefaultAdyen3DS2DelegateTest(
             defaultDispatcher = dispatcher,
             embeddedRequestorAppUrl = "embeddedRequestorAppUrl",
             base64Encoder = base64Encoder,
+            delegatedAuthentication = DelegatedAuthentication(
+                SavedStateHandle(),
+                configuration.merchantLogo
+            ),
             application = Application(),
         )
+        delegate.initialize(coroutineScope = TestScope())
```
Comment (OscarSpruit) [inline 3ds2/src/test/java/com/adyen/checkout/adyen3ds2/DefaultAdyen3DS2DelegateTest.kt:108]: This shouldn't be called from the before each

File: action/src/main/java/com/adyen/checkout/action/DefaultActionHandlingComponent.kt [hunk 8]
```diff
@@ -11,25 +11,42 @@ package com.adyen.checkout.action
...
 
+    private var isActionHandled: Boolean
+        get() = savedStateHandle[IS_ACTION_HANDLED] ?: false
+        set(value) {
+            savedStateHandle[IS_ACTION_HANDLED] = value
+        }
+
+    init {
+        // Restoring the state after process kill
+        if (isActionHandled) {
```
Comment (OscarSpruit) [inline action/src/main/java/com/adyen/checkout/action/DefaultActionHandlingComponent.kt:39]: Let's discuss this more if we really need this

File: bcmc/src/main/java/com/adyen/checkout/bcmc/BcmcComponent.kt [hunk 9]
```diff
@@ -46,7 +46,11 @@ class BcmcComponent internal constructor(
     ButtonComponent,
     ActionHandlingComponent by actionHandlingComponent {
 
-    override val delegate: ComponentDelegate get() = actionHandlingComponent.activeDelegate
+    override val delegate: ComponentDelegate
+        get() = when (val delegate = actionHandlingComponent.activeDelegate) {
```
Comment (OscarSpruit) [inline bcmc/src/main/java/com/adyen/checkout/bcmc/BcmcComponent.kt:50]: Why is this needed?

File: bcmc/src/main/java/com/adyen/checkout/bcmc/BcmcView.kt [hunk 10]
```diff
@@ -48,7 +48,9 @@ internal class BcmcView @JvmOverloads constructor(
     }
 
     override fun initView(delegate: ComponentDelegate, coroutineScope: CoroutineScope, localizedContext: Context) {
-        if (delegate !is BcmcDelegate) throw IllegalArgumentException("Unsupported delegate type")
+        // TODO (Auth) discuss. In case of restoring the state after process kill the view may receive
```
Comment (OscarSpruit) [inline bcmc/src/main/java/com/adyen/checkout/bcmc/BcmcView.kt:51]: Yeah let's discuss this 👍

File: bcmc/src/main/java/com/adyen/checkout/bcmc/DefaultBcmcDelegate.kt [hunk 11]
```diff
@@ -126,6 +133,46 @@ internal class DefaultBcmcDelegate(
...
+                )
+                null
+            } catch (e: IncompatibleClassChangeError) {
+                Logger.e(
+                    TAG,
+                    "delegatedAuthenticationData not set because " +
+                        "Authentication SDK version is incompatible with compiled version."
+                )
+                null
+            } catch (e: NullPointerException) {
```
Comment (OscarSpruit) [inline bcmc/src/main/java/com/adyen/checkout/bcmc/DefaultBcmcDelegate.kt:165]: Can't we make it work for tests? I don't like having code just for tests 😬

File: card/src/main/java/com/adyen/checkout/card/CardComponent.kt [hunk 12]
```diff
@@ -46,7 +46,11 @@ class CardComponent internal constructor(
     ButtonComponent,
     ActionHandlingComponent by actionHandlingComponent {
 
-    override val delegate: ComponentDelegate get() = actionHandlingComponent.activeDelegate
+    override val delegate: ComponentDelegate
+        get() = when (val delegate = actionHandlingComponent.activeDelegate) {
```
Comment (OscarSpruit) [inline card/src/main/java/com/adyen/checkout/card/CardComponent.kt:50]: Same here

File: card/src/main/java/com/adyen/checkout/card/StoredCardDelegate.kt [hunk 13]
```diff
@@ -171,6 +178,46 @@ internal class StoredCardDelegate(
         }
     }
 
+    @Suppress("TooGenericExceptionCaught")
```
Comment (OscarSpruit) [inline card/src/main/java/com/adyen/checkout/card/StoredCardDelegate.kt:181]: Why do we need to suppress this?

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/DelegatedAuthentication.kt [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    ): Boolean? {
+        val hasDACredential = try {
+            AdyenAuthentication.hasCredential(application, sdkInput)
+        } catch (e: ClassNotFoundException) {
+            Logger.e(TAG, "hasCredential not executed because Authentication SDK is not present in project.")
+            null
+        } catch (e: NoClassDefFoundError) {
+            Logger.e(TAG, "hasCredential not executed because Authentication SDK is not present in project.")
+            null
+        } catch (e: IncompatibleClassChangeError) {
```
[the remark this answers is not in the document]
Comment (tkuntubayev) [reply 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/DelegatedAuthentication.kt:87]: As merchants need to present Authentication SDK dependency in addition to Drop-In/Components SDK, it might be thrown in case if the merchant include Authentication SDK version which is incompatible with Drop-In/Components SDK version

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/DelegatedAuthentication.kt [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+            null
+        }
+        Logger.d(TAG, "hasDACredential: $hasDACredential")
+        return hasDACredential
+    }
+
+    internal fun startTimer(coroutineScope: CoroutineScope) {
+        isTimerStarted = true
+        publishTimerUpdate(currentMillisUntilFinished)
+        coroutineScope.launch(Dispatchers.Main) {
```
[the remark this answers is not in the document]
Comment (tkuntubayev) [reply 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/DelegatedAuthentication.kt:101]: CountDownTimer can be initialized only from UI thread, but startTimer called from non-UI

File: 3ds2/build.gradle [hunk 16]
```diff
@@ -41,13 +45,19 @@ dependencies {
     // Checkout
     api project(':ui-core')
 
+    // Delegated Authentication
+    // If Authentication SDK is present.
+    compileOnly libraries.adyenAuthentication
+
     // Dependencies
     api libraries.adyen3ds2
+    implementation libraries.material
```
[the remark this answers is not in the document]
Comment (tkuntubayev) [reply 3ds2/build.gradle:54]: `LinearProgressIndicator` view which is used to show timer progress is coming from material library. According to figma design the progress indicator has rounded corners and this view is specially designed for it (same approach is used for qr-code module, it also uses `LinearProgressIndicator`  ). There's an alternative to implement the logic of rounding `ProgressBar` on our side if we want to remove this dependency from 3DS2 component

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/DAAuthenticationResult.kt [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by temirlan on 8/12/2022.
+ */
+
+package com.adyen.checkout.adyen3ds2.model
+
+sealed class DAAuthenticationResult {
+    class AuthenticationSuccessful(val sdkOutput: String) : DAAuthenticationResult()
+    object NotNow : DAAuthenticationResult()
```
[the remark this answers is not in the document]
Comment (tkuntubayev) [reply 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/model/DAAuthenticationResult.kt:16]: Yes, it means that user clicked "Not now". Let's rename it to be more clear 👍

File: 3ds2/src/main/res/layout/view_da_authentication.xml (hunk 4 above)
Comment (tkuntubayev) [reply 3ds2/src/main/res/layout/view_da_authentication.xml:42]: The idea of this wrapper is to define percentage value from the width instead of setting some dp value. As for `QrCodeView` we use the second approach - let's use it here as well, I'll remove the wrapper

File: 3ds2/src/main/res/values/strings.xml (hunk 5 above)
Comment (tkuntubayev) [reply 3ds2/src/main/res/values/strings.xml:3]: We don't have translations yet. Can we request them?

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/ui/DAAuthenticationView.kt [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    private fun showRemoveCredentialsDialog(context: Context, onPositiveClick: () -> Unit) {
+        val onClickListener = DialogInterface.OnClickListener { dialog, which ->
+            when (which) {
+                DialogInterface.BUTTON_POSITIVE -> onPositiveClick.invoke()
+            }
+            dialog.dismiss()
+        }
+        MaterialAlertDialogBuilder(context, R.style.DAMaterialAlertDialogTheme)
```
[the remark this answers is not in the document]
Comment (tkuntubayev) [reply 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/ui/DAAuthenticationView.kt]: It was used to follow the figma design, however as we already use one type of dialog in the project - it will be better to save consistency, I'll replace it to `AlertDialog.Builder`

File: bcmc/src/main/java/com/adyen/checkout/bcmc/BcmcComponent.kt (hunk 9 above)
Comment (tkuntubayev) [reply bcmc/src/main/java/com/adyen/checkout/bcmc/BcmcComponent.kt:50]: This logic was used to get actual delegate before we started merging `viewFlow` using stateFlow. After merge with develop this logic is not needed anymore, so I'll remove it. Thanks for pointing

File: bcmc/src/main/java/com/adyen/checkout/bcmc/DefaultBcmcDelegate.kt (hunk 11 above)
Comment (tkuntubayev) [reply bcmc/src/main/java/com/adyen/checkout/bcmc/DefaultBcmcDelegate.kt:165]: Spend some time and found a solution. I've mocked the retrieving part in the tests and removed this catch 👍

File: card/src/main/java/com/adyen/checkout/card/StoredCardDelegate.kt (hunk 13 above)
Comment (tkuntubayev) [reply card/src/main/java/com/adyen/checkout/card/StoredCardDelegate.kt:181]: Was related to NullPointerException, removed it as it's not thrown anymore

File: 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/DelegatedAuthentication.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (OscarSpruit) [reply 3ds2/src/main/java/com/adyen/checkout/adyen3ds2/DelegatedAuthentication.kt:73]: Oh okay I see 👍

File: 3ds2/build.gradle (hunk 16 above)
[the remark this answers is not in the document]
Comment (OscarSpruit) [reply 3ds2/build.gradle:54]: No need. I was just double checking if we really need it
